### PR TITLE
Add support for cross-sample type import and update via file

### PR DESCRIPTION
**Pull request by labkey-susanh**
#### Rationale
Often users will have a single manifest for a number of samples of different types. Many of the fields for samples are shared, but some are distinct per sample type.  Currently, we require that users divide the manifests into one file per sample type in order to import them. This work adds support from the server to do that division of the file as part of the import process, when we are told that the file is a cross-sample-type file.  Later work will adjust the UI to allow for this.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-premium/pull/96
* https://github.com/LabKey/sampleManagement/pull/1858
* https://github.com/LabKey/biologics/pull/2151
* https://github.com/LabKey/inventory/pull/868

#### Changes
* Add new `crossSampleType` parameter understood by the import actions
* Add new `MultiSampleTypeDataIterator` to do the work of splitting the given file into multiple files and processing each of these files in turn

### CODE REVIEW

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java [hunk 1]
```diff
@@ -4084,21 +4087,58 @@ public class ImportSamplesAction extends AbstractExpDataImportAction
         public void validateForm(QueryForm queryForm, Errors errors)
         {
             _form = queryForm;
-            _form.setSchemaName("samples");
             _insertOption = queryForm.getInsertOption();
+            _crossTypeImport = Boolean.valueOf(getParam(Params.crossTypeImport));
+            if (!_crossTypeImport)
+                _form.setSchemaName("samples");
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: if crossTypeImport with a provided queryName, should we throw error?

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java [hunk 2]
```diff
@@ -4084,21 +4087,58 @@ public class ImportSamplesAction extends AbstractExpDataImportAction
...
+        @Override
+        protected void initRequest(QueryForm form) throws ServletException
+        {
+            QueryDefinition query = form.getQueryDef();
+            List<QueryException> qpe = new ArrayList<>();
+            TableInfo t = query.getTable(form.getSchema(), qpe, true);
+
+            if (!qpe.isEmpty())
+                throw qpe.get(0);
+            if (!_crossTypeImport && null != t)
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: can we have getCrossTypeImport() as a method on AbstractExpDataImportAction, which returns false by default? That way we don't need to override initRequest with such duplications

Comment (XingY) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: Also, is merge and update actually not supported for crossTypeImport?

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java [hunk 3]
```diff
@@ -4084,21 +4087,58 @@ public class ImportSamplesAction extends AbstractExpDataImportAction
...
+            }
+
+            _auditBehaviorType = form.getAuditBehavior();
+        }
+
+        @Override
+        protected int importData(DataLoader dl, FileStream file, String originalName, BatchValidationException errors, @Nullable AuditBehaviorType auditBehaviorType, TransactionAuditProvider.@Nullable TransactionAuditEvent auditEvent) throws IOException
+        {
+            TableInfo tInfo = _crossTypeImport ? new ExpMaterialTableImpl(ExpSchema.TableType.Materials.name(), new SamplesSchema(getUser(), getContainer()), ContainerFilter.current(getContainer())) : _target;
+            QueryUpdateService updateService = _crossTypeImport ? tInfo.getUpdateService() : _updateService;
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: Why is this needed? I would think _updateService should be set to the correct one on setTarget call above.

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java [hunk 4]
```diff
@@ -4084,21 +4087,58 @@ public class ImportSamplesAction extends AbstractExpDataImportAction
...
+        {
+            QueryDefinition query = form.getQueryDef();
+            List<QueryException> qpe = new ArrayList<>();
+            TableInfo t = query.getTable(form.getSchema(), qpe, true);
+
+            if (!qpe.isEmpty())
+                throw qpe.get(0);
+            if (!_crossTypeImport && null != t)
+            {
+                setTarget(t);
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: I think target tableInfo should be set for _crossTypeImport as well.

File: api/src/org/labkey/api/query/AbstractQueryUpdateService.java [hunk 5]
```diff
@@ -317,7 +317,7 @@ protected int _importRowsUsingInsertRows(User user, Container container, DataIte
 
     protected int _importRowsUsingDIB(User user, Container container, DataIteratorBuilder in, @Nullable final ArrayList<Map<String, Object>> outputRows, DataIteratorContext context, @Nullable Map<String, Object> extraScriptContext)
     {
-        if (!hasPermission(user, context.getInsertOption().updateOnly ? UpdatePermission.class : InsertPermission.class))
+        if (!context.isCrossTypeImport() && !hasPermission(user, context.getInsertOption().updateOnly ? UpdatePermission.class : InsertPermission.class))
```
Comment (XingY) [inline api/src/org/labkey/api/query/AbstractQueryUpdateService.java]: even with crossTypeImport, user should still need update/insert permission

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 6]
```diff
@@ -318,7 +317,9 @@ else if (nameObj instanceof Number)
     @Override
     public DataIteratorBuilder createImportDIB(User user, Container container, DataIteratorBuilder data, DataIteratorContext context)
     {
-        assert _sampleType != null : "SampleType required for insert/update, but not required for read/delete";
+        assert context.isCrossTypeImport() || _sampleType != null : "SampleType required for insert/update, but not required for read/delete";
+        if (context.isCrossTypeImport())
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:343]: Can the override be at loadRows level? It doesn't seem that all those extra handling associated with standard loadRows are needed so it might be cleaner to override there.

File: api/src/org/labkey/api/query/AbstractQueryImportAction.java [hunk 7]
```diff
@@ -796,7 +798,7 @@ public static int importData(DataLoader dl, TableInfo target, QueryUpdateService
                     throw new RuntimeSQLException(x);
             }
         }
-        else
+        else if (!context.isCrossTypeImport())
```
Comment (XingY) [inline api/src/org/labkey/api/query/AbstractQueryImportAction.java:816]: In the case of samples cross type, the target should also be not null, it should be the exp.material table.

File: experiment/src/org/labkey/experiment/ExpDataIterators.java [hunk 8]
```diff
@@ -2250,4 +2263,325 @@ else if (isMergeOrUpdate)
...
+                        if (updateService == null)
+                        {
+                            _context.getErrors().addRowError(new ValidationException("No update service available for sample type " + typeData.sampleType.getName() + "'."));
+                        }
+                        else
+                        {
+                            try (DataLoader loader = DataLoader.get().createLoader(typeData.dataFile, "text/plain", true, null, null))
+                            {
+                                // We do not need to configure the loader for renamed fields as that has been taken care of when writing the file.
+                                _context.setCrossTypeImport(false);
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/ExpDataIterators.java]: this can be moved outside of loop and only call once

File: experiment/src/org/labkey/experiment/ExpDataIterators.java [hunk 9]
```diff
@@ -2250,4 +2263,325 @@ else if (isMergeOrUpdate)
...
+        {
+            boolean hasNext = super.next();
+
+            if (_context.getErrors().hasErrors())
+                return hasNext;
+
+            if (!hasNext)
+            {
+                Collection<String> importOrderKeys = getImportOrderKeys();
+                if (!_context.getErrors().hasErrors())
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/ExpDataIterators.java:2331]: I think !_context.getErrors().hasErrors() is always false here

File: experiment/src/org/labkey/experiment/ExpDataIterators.java [hunk 10]
```diff
@@ -2250,4 +2263,325 @@ else if (isMergeOrUpdate)
...
+            if (_fileNameColIndex == null)
+                _context.getErrors().addRowError(new ValidationException("Sample type cannot be determined. Provide either a '" + StringUtils.join(SAMPLE_TYPE_FIELD_NAMES, "' or '") + "' column in the data."));
+        }
+
+        @Override
+        public boolean next() throws BatchValidationException
+        {
+            boolean hasNext = super.next();
+
+            if (_context.getErrors().hasErrors())
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/ExpDataIterators.java:2325]: If any error, typeData files should be deleted

File: experiment/src/org/labkey/experiment/ExpDataIterators.java [hunk 11]
```diff
@@ -2250,4 +2263,325 @@ else if (isMergeOrUpdate)
...
+                                _context.setCrossTypeImport(false);
+                                updateService.loadRows(_user, _container, loader, _context, null);
+                            }
+                            catch (SQLException | IOException e)
+                            {
+                                throw new RuntimeException(e);
+                            }
+                            finally
+                            {
+                                typeData.dataFile.delete();
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/ExpDataIterators.java]: These files only gets deleted if no other things went wrong. It's better to override close() and have close delete the files.

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java (hunk 1 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: Doesn't seem necessary really, but I could be convinced otherwise.

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java (hunk 2 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: Yes, good idea.

File: api/src/org/labkey/api/query/AbstractQueryUpdateService.java (hunk 5 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/query/AbstractQueryUpdateService.java]: Yes, but that's checked with the individual imports. No one has insert or update permission on exp.materials, so this fails, even if we set the target table to that table.

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java (hunk 6 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:343]: We already override loadRows in this class, and we do want some of that processing to happen when it's a cross-type import (e.g., handling the renaming of fields).  Since the thing we are changing is the data iterator, the override here seemed more natural, but I'm happy to discuss.

File: api/src/org/labkey/api/query/AbstractQueryImportAction.java (hunk 7 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/query/AbstractQueryImportAction.java:816]: For the cross-type import, the target table doesn't do much for us since we aren't actually inserting into `exp.materials`, or whatever table we choose.  All that's determined once we have read the file to pick out the types.

File: experiment/src/org/labkey/experiment/ExpDataIterators.java (hunk 8 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/ExpDataIterators.java]: done.

File: experiment/src/org/labkey/experiment/ExpDataIterators.java (hunk 9 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/ExpDataIterators.java:2331]: Not always. We can set an error if we find a cycle in the dependencies while doing an insert.

File: experiment/src/org/labkey/experiment/ExpDataIterators.java (hunk 11 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/ExpDataIterators.java]: Ah, didn't know there was a `close` method. That's definitely better.

File: api/src/org/labkey/api/query/AbstractQueryUpdateService.java (hunk 5 above)
Comment (XingY) [reply api/src/org/labkey/api/query/AbstractQueryUpdateService.java]: If that's the case, then the override should be limited to SampleQUS, not abstractQUS.

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java (hunk 2 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: Didn't actually need to override anything since `_crossTypeImport` is defined in the base class.

Yes, merge and update are meant to be supported. Is there something here that suggests they aren't being fully supported?

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java (hunk 4 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: It doesn't really work out well for checking permissions. The methods `canUpdate`, `canInsert`, etc. already handle the case of `_target` being null, checking the proper permissions based on the container, which is the most we can hope for as a starting point for a cross-type import.

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java (hunk 3 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: See comment above.

File: experiment/src/org/labkey/experiment/ExpDataIterators.java (hunk 10 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/ExpDataIterators.java:2325]: good point. This should now get taken care of by the `close` method.

File: api/src/org/labkey/api/query/AbstractQueryUpdateService.java (hunk 5 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/query/AbstractQueryUpdateService.java]: I think it's safe either way with the current implementation since `_target` will be null when `crossTypeImport` is true so users won't get very far even if trying to bypass the permission check by setting that parameter, but you get a better error message if it's a permissions failure, so I added a method that's overridden in `SampleTyepUpdateServiceDI`.

**Review by XingY — APPROVED** [review]
